### DISCUSSION TITLE
Struggling with Paths to match

### DISCUSSION THREAD

**Discussion by castortech** [dnGrep/dnGrep#511] [Q&A] [answered]
I am having a hard time specifying valid paths to match. My files have dots in them and here I want to search within jar files that are matching the pattern of "org.eclipse.*.jar" (all jar files starting with org.eclipse).

As shown in the attached image (search for regex . to match just any file after) shows matching of bcpkix.source....jar.

What am I missing.
![Paths to match issue](https://user-images.githubusercontent.com/85858/120889777-b3793f00-c5cc-11eb-83bd-bee117bc3746.PNG)

**Comment by doug24** [comment, marked as answer]
I'm not sure you can do what you want.  dnGrep treats archives as compressed directories, not as files.  So by checking "Search in archives", dnGrep will search all the archive types as specified in the Options.  You don't specify the archive type or name in the "Paths to match", that is where you set the file name/file type filter for the type of file to search, which may or may not be in an archive.   Yes, "Paths to match" should probably be "Files to match".

dnGrep doesn't have a way to specify an include-only path like you want for org.eclipse.   But the "Paths to ignore" does allow you exclude paths by name or pattern, and that applies to parts of the path both outside and inside the archive.  So you would have to ignore everything that is not org.eclipse.  Both Path to match and Paths to ignore accept comma or semi-colon separated sets of patterns.

**Reply by castortech** [reply]
Doug,
Thanks for the quick reply. 

Ok that explains part of it. But why are we seeing in my image that it is matching files inside the archive that clearly don't match the pattern?

**Reply by doug24** [reply]
That is a good question, and I don't know.  It doesn't look like it should have matched and in the simple test case I just set up, it doesn't do what you are seeing.

**Reply by doug24** [reply]
After looking again, I do see why it is matching now.   When using the regular expression file filter (Paths that match), the file iterator code returns every file as a full path to filtering code, and the regex is applied to the full path.    So for example, the first file would get returned like this:
`C:\eclipse\baws2003\.metadata.plugins\org.eclipse.pde.core\.bundle_pool\plugins\bcpkix.source_1.68.0.jar///META-INF\MANIFEST.MF`
and your filter will match the substring:
`org.eclipse.pde.core\.bundle_pool\plugins\bcpkix.source_1.68.0.jar`
and that filter will match every file in every jar under org.eclipse.pde.core

Note the /// is a convention used in dnGrep to represent paths within archives.

So maybe you can do what you want by changing your filter to something like
`org\.eclipse[^\\]*\.jar`
to stop the regex from matching names in deeper subdirectories.

But you are still going to get every file inside the org.eclipse jar files - and maybe that's what you want.  But if not, maybe you can continue the regular expression to add additional filtering past the `.jar`

**Reply by castortech** [reply]
Thanks, great explanation and it all works now. BTW, I was only trying to get every file to test the file pattern matching.
